src/storage/src/source/postgres.rs

```diff
@@ -370,272 +627,211 @@ impl PostgresSourceReader {
-                            if !self.source_tables.contains_key(&rel_id) {
-                                continue;
-                            }
-                            let new_tuple = insert.tuple().tuple_data();
-                            let row = try_fatal!(self.row_from_tuple(rel_id, new_tuple));
-                            inserts.push(row);
-                        }
-                        Update(update) => {
-                            self.metrics.updates.inc();
-                            let rel_id = update.rel_id();
-                            if !self.source_tables.contains_key(&rel_id) {
-                                continue;
-                            }
-                            let old_tuple = try_fatal!(update
-                                .old_tuple()
-                                .ok_or_else(|| anyhow!("Old row missing from replication stream for table with OID = {}. \
-                                        Did you forget to set REPLICA IDENTITY to FULL for your table?", rel_id)))
-                            .tuple_data();
-                            let old_row = try_fatal!(self.row_from_tuple(rel_id, old_tuple));
-                            deletes.push(old_row);
-
-                            // If the new tuple contains unchanged toast values, reuse the ones
-                            // from the old tuple
-                            let new_tuple = update
-                                .new_tuple()
-                                .tuple_data()
-                                .iter()
-                                .zip(old_tuple.iter())
-                                .map(|(new, old)| match new {
-                                    TupleData::UnchangedToast => old,
-                                    _ => new,
-                                });
-                            let new_row = try_fatal!(self.row_from_tuple(rel_id, new_tuple));
-                            inserts.push(new_row);
-                        }
-                        Delete(delete) => {
-                            self.metrics.deletes.inc();
-                            let rel_id = delete.rel_id();
-                            if !self.source_tables.contains_key(&rel_id) {
-                                continue;
-                            }
-                            let old_tuple = try_fatal!(delete
-                                .old_tuple()
-                                .ok_or_else(|| anyhow!("Old row missing from replication stream for table with OID = {}. \
-                                        Did you forget to set REPLICA IDENTITY to FULL for your table?", rel_id)))
-                            .tuple_data();
-                            let row = try_fatal!(self.row_from_tuple(rel_id, old_tuple));
-                            deletes.push(row);
+        let closer = self.sender.clone();
+        loop {
+            tokio::select! {
```

Comment:
mind adding a comment like "yo these are both cancel-safe"? (so long as that fits into the grander plan for `tokio::select`)

Reply:
oh interesting its more subtle than that...but still safe I thini...luckily we can just move to `SourceReader::next` in the future and avoid this altogether